runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientConnectionPool.java

```diff
@@ -738,21 +735,47 @@ private void doStreamData(
             replySeq = sequence;
             replyAck = acknowledge;
 
-            doData(sender, originId, routedId, replyId, replySeq, replyAck, replyMax,
-                traceId, authorization, flags, replyBud, reserved, payload, offset, length, extension);
+            if (responseBytes == 0)
+            {
+                nexResponseId++;
+                final ResponseHeaderFW responseHeader = responseHeaderRO.wrap(payload, offset, offset + length);
+                responseBytes = responseHeader.length() + KAFKA_FRAME_LENGTH_FIELD_OFFSET;
+            }
+
+            responseBytes -= length;
+
+            if (!KafkaState.replyClosing(state))
+            {
+                doData(sender, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                    traceId, authorization, flags, replyBud, reserved, payload, offset, length, extension);
+            }
+            else
+            {
+                if (responseBytes == 0)
+                {
+                    doStreamEnd(traceId);
+                }
+            }
         }
 
         private void doStreamEnd(
             long traceId)
         {
             if (!KafkaState.replyClosed(state))
             {
-                state = KafkaState.closedReply(state);
+                if (nextRequestId == nexResponseId)
+                {
+                    state = KafkaState.closedReply(state);
 
-                doEnd(sender, originId, routedId, replyId, 0, 0, 0,
-                    traceId, authorization, EMPTY_EXTENSION);
+                    doEnd(sender, originId, routedId, replyId, 0, 0, 0,
+                        traceId, authorization, EMPTY_EXTENSION);
 
-                streamsByInitialIds.remove(initialId);
+                    streamsByInitialIds.remove(initialId);
+                }
+                else
+                {
+                    state = KafkaState.closingReply(state);
```

Comment:
Please move this above `if` statement, so we are always `closing` and then conditionally also `closed`.